### DISCUSSION TITLE
Static typing for unique name nodes and node paths (find_child, get_node)

### DISCUSSION THREAD

**Discussion by FloBec** [godotengine/godot-proposals#9380] [Scripting]
*not a trained programmer*
To make my code more readable, i try to use single lines to set properties of UI nodes.
Unfortunately these can not be typed in the same single line. Creating new variables every time would extend my code vertically and thus reduce readability.

Current code examples:
```python
newMutator.find_child("mutatorProperty").text = str(mutator.text)
%itemDescription.text = item_data.description
``` 
result: "The property "text" is not present on the inferred type "Node" (but may be present on a subtype). <GDScript Error>UNSAFE_PROPERTY_ACCESS"
 
Proposal:
```python
newMutator.find_child("mutatorProperty").text: String= str(mutator.text)
%itemDescription.text: String = item_data.description
```

**Comment by Obscurite1221** [comment]
First of all, vertically decompressing your code is good in most cases, as it is usually more readable, harder to lose your place when reading, and ideally all optimized out anyways. Secondly, this is more an issue of programming habit. GDScript is dynamically typed, so if you want static typed external access, you should use getters and setters. 
E.g.
newMutator.find_child("mutatorProperty").set_text(mutator.text)
and inside the mutatorProperty,
func set_text(variableName : String) -> void:
(tab)text = variableName
(tab)pass

This tells Godot that it's a function that should only accept a variable containing a string (which is known at runtime) and not to expect anything back in return. 
Also, you already had a fix for the second line in your provided example. 
%itemDescription.text = item_data.description
->
%itemDescription.text = str(item_data.description)

Although realistically, GDScript shouldn't really care what you're assigning to what, so I don't know why that would error. You can assign a string to an int and vice versa, so idk what the issue is there.

**Reply by FloBec** [reply]
Creating a setter function for every label text may be a bit much, so maybe i will look at creating one function that also takes the target variable as an argument. 

As for context: I came across two posts on how to improve performance in a godot project.
- https://www.beep.blog/2024-02-14-gdscript-typing/
- https://allenwp.com/blog/2023/10/03/how-to-enforce-static-typing-in-gdscript/
Following these i enabled the warnings and amended my code. I haven't measured the performance impact. For me its a learning exercise and a way to improve my general coding.

As for the second example: The warning is that the unique node type is unknown, not that a string cannot be assigned.
To make the node type explicit, maybe one of these makes sense.
```python
%itemDescription.text: Label = str(item_data.description)
%itemDescription.text as Label = str(item_data.description)
``` 

Its not that the current code, or the setter-approach would not work perfectly fine in my project - i am merely trying to explore ideas for a  concise syntax.

**Comment by theraot** [comment]
Any reason why disabling the warning `UNSAFE_PROPERTY_ACCESS` is not a viable solution? It is in Project Settings under `debug/gdscript/warnings/unsafe_property_access` and set to Ignore by default. So I presume you are enabled for a reason.

Addendum: Consider that in the proposed solution `%itemDescription.text: String = item_data.description` you would be telling Godot the type, which might or might not be the correct type. That is akin to suppressing the warning (i.e. you would be telling Godot to treat this code as correct and not check).

**Comment by dalexeev** [comment]
You can use type casting `(%Label as Label).text`, `@warning_ignore`, or an `@onready` variable.

**Reply by FloBec** [reply]
neat, this also enables auto-complete. thanks.
```python
(newMutator.find_child("mutatorProperty") as Label).text = str(mutator.text)
```
